### PR TITLE
#132: Support composed members for non-contained ComposedTypeRef

### PR DESCRIPTION
**Pull request by mor-n4**
see #132

Tests will be added by @qtran-n4 as part of #73, because noticing the problem fixed by this PR on the surface is easier with functionality from #73.

### CODE REVIEW

**Comment by mor-n4** [comment]
@qtran-n4 builds still running, but you can already start looking through the changes.

File: plugins/org.eclipse.n4js.model/emf-gen/org/eclipse/n4js/n4JS/MemberAccess.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright (c) 2016 NumberFour AG.
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js.model/emf-gen/org/eclipse/n4js/n4JS/MemberAccess.java]: 2017

Comment (mor-n4) [inline plugins/org.eclipse.n4js.model/emf-gen/org/eclipse/n4js/n4JS/MemberAccess.java]: this is EMF-generated code; we can't bump the copyright year for individual EMF entities

File: plugins/org.eclipse.n4js.model/emf-gen/org/eclipse/n4js/n4JS/IndexedAccessExpression.java [hunk 2]
```diff
@@ -28,7 +28,7 @@
  * @model
  * @generated
  */
-public interface IndexedAccessExpression extends Expression {
+public interface IndexedAccessExpression extends Expression, MemberAccess {
```
Comment (mor-n4) [inline plugins/org.eclipse.n4js.model/emf-gen/org/eclipse/n4js/n4JS/IndexedAccessExpression.java]: excellent find!!

But: we cannot solve this for all cases, because we have a conflict of goals here. We cannot say:
* I want the scoping to return members that are contained in a resource AND
* I want to invoke getAllElements() on the scope AND
* I don't want the composed member cache to explode.

There's no way to solve all of that simultaneously.

But I'll try to improve the client code handling index expressions to avoid calling #getAllElements() + I'll update API doc.

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics [hunk 3]
```diff
@@ -695,9 +698,10 @@ from {
 			val staticAccess = (targetTypeRef instanceof TypeTypeRef)
 			val checkVisibility = false // access modifiers checked in validation
 			val scope = memberScopingHelper.createMemberScope(targetTypeRef, expr, checkVisibility, staticAccess)
-			val member = memberScopingHelper.findUniqueMemberForName(scope, memberName, staticAccess)
-			if(member !== null) {
-				G |- member : var TypeRef memberTypeRef
+			val memberDesc = scope.getSingleElement(qualifiedNameConverter.toQualifiedName(memberName));
```
Comment (qtran-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics]: The original method `findUniqueMemberForName` filters `IEObjectDescriptionWithError`. In this refactored version, shouldn't we check `! (memberDesc instanceof IEObjectDescriptionWithError)` as well?

File: tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/scoping/ComposedMemberCachingTest.xtend [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+import org.junit.Test
+
+
+/**
+ * Tests behavior of member scoping related to the caching of composed members. For further information, see
+ * API documentation of methods:
+ * {@link MemberScopingHelper#createMemberScope(TypeRef, MemberAccess, boolean, boolean)} and
+ * {@link MemberScopingHelper#createMemberScopeAllowingNonContainedMembers(TypeRef, EObject, boolean, boolean)}.
+ */
+class ComposedMemberCachingTest extends AbstractN4JSTest {
```
Comment (qtran-n4) [inline tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/scoping/ComposedMemberCachingTest.xtend]: Really nice tests!!

File: tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/scoping/ComposedMemberCachingTest.xtend [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	@Test
+	def void testContainedMemberWhenUsingNonContainedTypeRef() {
+		val script = snippet.parse; // do not validate!
+		val resource = script.eResource as N4JSResource;
+
+		val varDecl = script.eAllContents.filter(VariableDeclaration).head;
+		val propAccess = script.eAllContents.filter(ParameterizedPropertyAccessExpression).head;
+
+		// use a receiver type reference that is *NOT* contained in a resource
+		val receiverTypeRef = TypeUtils.copy(varDecl.declaredTypeRef);
```
Comment (qtran-n4) [inline tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/scoping/ComposedMemberCachingTest.xtend]: Should we assert

```
assertNull(receiverTypeRef, receiverTypeRef.eResource) 
```

to be absolutely sure?

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics (hunk 3 above)
Comment (mor-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesystem/n4js.xsemantics]: done

File: tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/scoping/ComposedMemberCachingTest.xtend (hunk 5 above)
Comment (mor-n4) [inline tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/scoping/ComposedMemberCachingTest.xtend]: the copy is never contained, so I don't think we need this

Comment (mor-n4) [inline tests/org.eclipse.n4js.lang.tests/src/org/eclipse/n4js/tests/scoping/ComposedMemberCachingTest.xtend]: well, it won't hurt and I have to make another commit anyway, so I added that assertion

done.

**Comment by mor-n4** [comment]
builds are green (finally)
merging now ...

**Comment by qtran-n4** [comment]
In the merged solution, we differentiate between two modi:

- The client code of the scoping requires that a composed member must be contained in a resource.
- The client code of the scoping does NOT require that a composed member must be contained in a resource. This is usually the case when the client needs to use the scoping infrastructure for say validation (for example `N4JSExpressionValidator#internalCheckCtorVisibility`).

I am thinking about the question: is the second case a misuse of scoping?

What is the purpose of scoping? Scoping is used by the linking service to wire AST/TModule elements with each other. Validation should happen after the scoping is done and thus can use the information provided by scoping during the validation. That scoping is CONSTRUCTED/CREATED during validation seems weird to me... 

If the creation of scope is not allowed during validation etc., the second mode may become obsolete...

**Comment by mor-n4** [comment]
Mode 2 is used only in cases where no ordinary scoping takes place (otherwise the result from the ordinary scoping could be used, as you said). Examples: typing & validation of new expressions (for looking up the ctor), typing & validation of destructuring patterns.

In addition, mode 2 is also required if, for w/e reason, methods like `#getAllElements()` have to be used and excessive caching of composed members is to be avoided. Afaik, ordinary scoping triggered by Xtext will not use methods like `#getAllElements()` and is thus ok with mode 1.
